### ISSUE TITLE
conda-forge-linter Meta-Mega-Issue

### ISSUE THREAD

**Issue by leycec** [conda-forge/conda-smithy#736]
Per @CJ-Wright's [sagacious advice](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-381300150) at [my first conda-forge recipe](https://github.com/conda-forge/staged-recipes/pull/5574), this large-scale issue collects various interrelated ~~critiques~~ concerns of current conda-forge-linter behaviour.

Yes, I probably should have created one distinct issue per concern. Sadly, I am lazy and sleep-deprived. I hope the current approach suffices without badly annoying anyone.

Before we begin, I'd like to thank @carlodri, @CJ-Wright, and @jakirkham for their tireless efforts in seeing my first pull request through to its ultimately happy conclusion. This meta-mega-issue's for you.

## What's So Wrong with conda-forge-linter, Anyway?

Nothing, assuming you enjoy feasting on insanity. I don't. Unless I'm playing Bloodborne. In which case I do.

Sadly, conda-forge linter is not Bloodborne. It's clearly... a work in progress. That's understandable. Given scarce developer resources, I'd throw everything I have at migrating to conda-build 3, too. Still, the linter sufficiently [threw everyone for a loop at my pull request](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-379522487) that it arguably squandered more developer time than it saved. That's bad. Automation is supposed to save time – not recklessly consume even *more* of it.

When resources permit,<sup>*...maybe after conda-build 3? maybe never?*</sup> it'd be awe-inspiring if you could throw a sacrificial developer or two at the following linter issues.

### Silent Failures Everywhere

The linter appears to silently fail after the first failure. At the first failure, the linter helpfully emitted [an error message](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-378863222). At each subsequent failure, the linter unhelpfully emitted... *nothing.* In fact, [nobody even knew whether the linter was being re-run or not](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-379434605).

Instead, I pompously propose the following:

> Refactor the linter to **append one new comment to the current pull request after each (re-)linting of this request** (e.g., after a new commit triggers a github webhook triggering the linter). Even if that comment is the exact same as the prior comment that it appended for the prior linting, appending a new comment demonstrates to everyone involved that the linter is actually doing something.

That's vital.

### Meaningless Errors Everywhere

The linter doesn't appear to emit meaningful errors. At the first failure, the linter unhelpfully emitted [the error message](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-378863222): "Failed to even lint the recipe (might be a conda-smithy bug) :cry: ".

In actuality, the linter at least partially linted the recipe before fatally choking on the conda-build 3-specific `requirements/host:` list. Presumably, the linter internally raised an exception. But neither this exception nor any related error was conveyed to the user in the above error message.

Instead, I pompously propose the following:

> Refactor the linter to **embed exception messages and/or tracebacks in the comments that it appends to pull requests.** Exception messages would be equally meaningful for both conda-forge users and maintainers. Tracebacks would be utterly meaningless to users, of course, but probably invaluable to maintainers.

In either case, something is better than nothing – and nothing is what we currently got.

### Draconian Strictness Everywhere (*Maybe*)

The linter appears to be overly draconian. Like, [Soup Nazi draconian](https://www.youtube.com/watch?v=M2lfZg-apSA).

On the one hand, it's a linter; it's *supposed* to be draconian. On the other hand, it appears to silently reject otherwise valid input that fails to conform to the exact example template for conda-forge `meta.yaml` files. Maybe? Nobody was quite sure about what was or was not permissible. [Things got a bit... confusing](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-379167097). The prior two issues probably didn't help.

I have no meaningful suggestions here, mostly because I'm unclear how strict (or not) the linter actually is. This stale rice ball adequately summarizes my confusion. :rice_ball: 

### @conda-forge-admin, Lint or Die!

The `@conda-forge-admin, please lint` command [didn't appear to do anything](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-378868960). I suspect this is because that command only applies to pull requests in existing feedstocks, whereas mine was a pull request creating a new feedstock. This might also be a sub-issue of the first point above.

In either case, it really didn't help.

## Is That It, Already?

That's it, because I am even more sleep-deprived now. Canadian winter will do that to a human.

To summarize the current state of affairs: conda-forge-linter need work badly. Thanks again, all!

**Comment by isuruf** [comment]
> Instead, I pompously propose the following: Refactor the linter to append one new comment to the current pull request after each (re-)linting of this request.

For me linter pushing a message everytime I push a commit is annoying. You can check whether the linter was run on a specific commit by clicking the CI status (A red x or a green tick on the right of the commit sha). That should give you the status and clicking on details should give you the relevant older message.

> Refactor the linter to embed exception messages and/or tracebacks in the comments that it appends to pull requests.

I'm not sure, but tracebacks are not given because it may expose secret tokens. But we should mention that running `conda smithy recipe-lint recipe/` locally should give the error message with a traceback.

> The linter doesn't appear to emit meaningful errors

The new version 3.0.0.rc1 has better meaningful errors when there's unexpected sections.

> The @conda-forge-admin, please lint command didn't appear to do anything.

This was an issue with the github hook where it was not sending the correct events. I've fixed this now.

Please let me know if I've missed anything from your post.

**Comment by dougalsutherland** [comment]
I think it would be nice to get tracebacks when the linter crashes. What situation would that expose a secret key in?

If not, agree that it should definitely include a message about how to lint locally.

**Comment by jakirkham** [comment]
Thanks for the feedback, @leycec. Agree the linter has room to grow.

The message you have seen from the linter comes up when it has an internal exception. We don't currently publish what those are. However expect we would be able to fix a lot of issues more quickly if we started doing that. We have to figure out how to do this safely, but it seems very worthwhile to do. So this feedback is very helpful.

As to the linter commenting everytime, we use to have this, but it was turned off due to noise. Though one should be able to explicitly ask for linting to occur by writing `@conda-forge-admin, please relint` in a comment on the PR. Maybe this would help if it is unclear whether the linter changed. (side note: we should document this. :)

Thanks for adding the note about linting locally as well.
